compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -455,16 +455,17 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         }
 
         ensure_sufficient_stack(|| {
-            match obligation.predicate.skip_binders() {
+            let bound_predicate = obligation.predicate.bound_atom(self.infcx().tcx);
```

Comment:
using `bound_atom` shifts escaping bound vars outwards which should not happen here.

How about adding a new method `bound_atom_with_opt_escaping` and instead asserting `Predicate::Atom(pred) => `debug_assert!(!pred.has_escaping_bound_vars())` in `bound_atom`

Reply:
Ah, fair enough. The `Atom` condition would be caught below (at least in the `PredicateAtom::Trait` case.

This is I guess one of the cases that is exposed as being a little "weird" when being more explicit about the binding. In the most strict sense, if this actually *does* care about the binders being potentially shifted, there is some information information "escaping" the `Binder` encapsulation.

I like having `bound_atom` *not* be able to return with escaping bound vars. But I don't think we need a new method for the case where it might. Unless I'm missing something, `bound_atom` is only used in three places currently (and nowhere *important*), so I don't think we're relying on this. So I'm gonna switch that and check tests.

Reply:
there are 2 places which currently have to deal with escaping vars, so for these we actually have to shift the escaping vars